python/ucxx/ucxx/_lib/libucxx.pyx

```diff
@@ -95,9 +95,20 @@ def _get_host_buffer(uintptr_t recv_buffer_ptr):
     return np.asarray(HostBufferAdapter._from_host_buffer(host_buffer))
 
 
-cdef shared_ptr[Buffer] _rmm_am_allocator(size_t length):
-    cdef shared_ptr[RMMBuffer] rmm_buffer = make_shared[RMMBuffer](length)
-    return dynamic_pointer_cast[Buffer, RMMBuffer](rmm_buffer)
+# A pure C++ is used here for the allocator, a Cython function always raises
+# exceptions back to Python, but the allocator here is executed as a callback
+# directly in C++, therefore exceptions have to be raised to the C++ caller
+# directly.
+cdef extern from *:
```

Comment:
I think you can do this in cython by marking the function as `noexcept` (which doesn't mean "nothrow"!, https://cython.readthedocs.io/en/latest/src/userguide/nogil.html#marking-functions-as-able-to-run-without-the-gil), but this change is fine.

Reply:
This is indeed much better, I could have sworn this existed but couldn't find it for some reason, it seems I don't know how to use Cython's docs. Reverted to the previous plus `noexcept`/`nogil` in https://github.com/rapidsai/ucxx/pull/315/commits/bf54be9367a7d91ffe15a14bd0059954f223f6fa .